packages/darklang/languageTools/nameResolver.dark

```diff
@@ -0,0 +1,280 @@
+              )
+
+          | modules ->
+            // TODO : check if user program exists
+            Stdlib.Result.Result.Ok(
+              (ProgramTypes.FQName.UserProgram
+                { modules = modules
+                  name = (constructor name)
+                  version = 0L })
+              |> ProgramTypes.FQName.FQName.UserProgram
+            )
+
+      let namesToTry
+        (given: List<String>)
+        (modules: List<String>)
+        : List<List<String>> =
+        match modules with
+        | [] -> [ given ]
+        | _ ->
+          let rest = Stdlib.List.dropLast modules
+          let current = Stdlib.List.append modules given
+          Stdlib.List.append [ current ] (namesToTry given rest)
+
+      let resolve
+        (constructor: String -> 'name)
+        (parser: String -> Stdlib.Result.Result<String>)
+        (nameErrorType: LanguageTools.RuntimeErrors.NameResolution.NameType)
+        (packageThingExists: String -> Bool)
+        (builtinThingExists: String -> Bool)
+        (allowError: Bool)
+        (currentModule: List<String>)
+        (name: LanguageTools.WrittenTypes.Name)
+        : LanguageTools.ProgramTypes.NameResolution<LanguageTools.ProgramTypes.FQName.FQName<'name>> =
+
+        match name with
+        | Unresolved given ->
+          let result =
+            Stdlib.List.fold
+              (NameResolver.namesToTry given currentModule)
+
+              (Stdlib.Result.Result.Error(
+                LanguageTools.RuntimeErrors.NameResolution.Error
+                  { nameType = nameErrorType
+                    errorType =
+                      LanguageTools.RuntimeErrors.NameResolution.ErrorType.NotFound
+                    names = given }
+              ))
+
+              (fun currentResult pathToTry ->
+                match currentResult with
+                | Ok _ -> currentResult
+                | Error _ ->
+                  let newResult =
+                    resolveName
+                      constructor
+                      parser
+                      pathToTry
+                      packageThingExists
+                      builtinThingExists
+                      nameErrorType
+
+                  match newResult with
+                  | Ok _ -> newResult
+                  | Error _ -> currentResult)
+
+          match result with
+          | Ok result -> Stdlib.Result.Result.Ok result
+          | Error err ->
+            if allowError then
+              Stdlib.Result.Result.Error err
+            else
+              Stdlib.Result.Result.Error "TODO: Raise exception"
+
+
+        | _ -> Stdlib.Result.Result.Error "TODO : builtin and user program"
+
+
+
+      module TypeName =
+        let packageTypeExists (typeName: String) : Bool =
+
+          let typ =
+            Stdlib.HttpClient.request
+              "GET"
+              ("http://dark-packages.dlio.localhost:11003/type/by-name/" ++ typeName)
```

Comment:
I think it's worth extracting out a `LanguageTools.PackageManager` module that lives on its own, with this sort of usage contained within such. (`LanguageTools.PackageManager.Type.byName urlBase typeName` maybe)

That said, that can wait for another PR.

Reply:
extracting that will also given us an opportunity for some caching of those packages, which will be needed before long